### PR TITLE
  [TunableOp] Fix offline sub-matrix detection when a leading dimension aliases a GEMM dim

### LINKED ISSUE #5553: [Issue]: _create_matrices() failed to create correct matrices for TunableOp offline tuning
### Problem Description

I am trying to trigger a TunableOp search for GemmTunableOp_Half_NT(nt_65536_32768_384_ld_65536_65536_65536) very specifically, because this op seems to cause a ring comp_1.0.1 timeout for my PC (and hence desktop failure) after some time. However, when I try to replicate it with offline tuning, it couldn't create the matrices needed and instead searched for GemmTunableOp_Half_NT(nt_65536_32768_384_ld_65536_32768_65536).

### Operating System

Linux Mint 21.3 (Virginia)

### CPU

AMD Ryzen 5 5500

### GPU

AMD Radeon VII

### ROCm Version

ROCm 7.14.0a2026053

### ROCm Component

_No response_

### Steps to Reproduce

Follow in instructions in [RELEASES.md](https://github.com/ROCm/TheRock/issues/RELEASES.md) for Multi-arch releases, install multi-arch ROCm and PyTorch Python packages.
In this case, the specific one I installed is: rocm-7.14.0a20260531 and amd-torch-device-gfx906 2.11.0+rocm7.14.0a20260531

Create a csv file called "gemm_to_tune.csv"
Put this into it:
`GemmTunableOp_Half_NT,nt_65536_32768_384_ld_65536_65536_65536`

Then run the following Python script:
```
import torch
import os

os.environ["PYTORCH_TUNABLEOP_ENABLED"] = "1"
os.environ["PYTORCH_TUNABLEOP_TUNING"] = "1"
os.environ["PYTORCH_TUNABLEOP_VERBOSE"] = "3"
os.environ["PYTORCH_TUNABLEOP_HIPBLASLT_ENABLED"] = "0" # Have to be false for gfx906 or rocblaslt error
os.environ["PYTORCH_TUNABLEOP_FILENAME"] = "tuned_gemm.csv"

torch.cuda.tunable.tune_gemm_in_file('gemm_to_tune.csv')
```

### (Optional for Linux users) Output of /opt/rocm/bin/rocminfo --support

```
ROCk module version 6.16.13 is loaded
=====================    
HSA System Attributes    
=====================    
Runtime Version:         1.21
Runtime Ext Version:     1.21
System Timestamp Freq.:  1000.000000MHz
Sig. Max Wait Duration:  18446744073709551615 (0xFFFFFFFFFFFFFFFF) (timestamp count)
Machine Model:           LARGE                              
System Endianness:       LITTLE                             
Mwaitx:                  ENABLED
XNACK enabled:           NO
DMAbuf Support:          YES
VMM Support:             YES

==========               
HSA Agents               
==========               
*******                  
Agent 1                  
*******                  
  Name:                    AMD Ryzen 5 5500                   
  Uuid:                    CPU-XX                             
  Marketing Name:          AMD Ryzen 5 5500                   
  Vendor Name:             CPU                                
  Feature:                 None specified                     
  Profile:                 FULL_PROFILE                       
  Float Round Mode:        NEAR                               
  Max Queue Number:        0(0x0)                             
  Queue Min Size:          0(0x0)                             
  Queue Max Size:          0(0x0)                             
  Queue Type:              MULTI                              
  Node:                    0                                  
  Device Type:             CPU                                
  Cache Info:              
    L1:                      32768(0x8000) KB                   
  Chip ID:                 0(0x0)                             
  ASIC Revision:           0(0x0)                             
  Cacheline Size:          64(0x40)                           
  Max Clock Freq. (MHz):   4359                               
  BDFID:                   0                                  
  Internal Node ID:        0                                  
  Compute Unit:            12                                 
  SIMDs per CU:            0                                  
  Shader Engines:          0                                  
  Shader Arrs. per Eng.:   0                                  
  WatchPts on Addr. Ranges:1                                  
  Memory Properties:       
  Features:                None
  Pool Info:               
    Pool 1                   
      Segment:                 GLOBAL; FLAGS: FINE GRAINED        
      Size:                    65725116(0x3eae2bc) KB             
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Recommended Granule:4KB                                
      Alloc Alignment:         4KB                                
      Accessible by all:       TRUE                               
    Pool 2                   
      Segment:                 GLOBAL; FLAGS: EXTENDED FINE GRAINED
      Size:                    65725116(0x3eae2bc) KB             
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Recommended Granule:4KB                                
      Alloc Alignment:         4KB                                
      Accessible by all:       TRUE                               
    Pool 3                   
      Segment:                 GLOBAL; FLAGS: KERNARG, FINE GRAINED
      Size:                    65725116(0x3eae2bc) KB             
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Recommended Granule:4KB                                
      Alloc Alignment:         4KB                                
      Accessible by all:       TRUE                               
    Pool 4                   
      Segment:                 GLOBAL; FLAGS: COARSE GRAINED      
      Size:                    65725116(0x3eae2bc) KB             
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Recommended Granule:4KB                                
      Alloc Alignment:         4KB                                
      Accessible by all:       TRUE                               
  ISA Info:                
*******                  
Agent 2                  
*******                  
  Name:                    gfx906                             
  Uuid:                    GPU-d9ac292172da5f19               
  Marketing Name:          AMD Radeon VII                     
  Vendor Name:             AMD                                
  Feature:                 KERNEL_DISPATCH                    
  Profile:                 BASE_PROFILE                       
  Float Round Mode:        NEAR                               
  Max Queue Number:        128(0x80)                          
  Queue Min Size:          64(0x40)                           
  Queue Max Size:          131072(0x20000)                    
  Queue Type:              MULTI                              
  Node:                    1                                  
  Device Type:             GPU                                
  Cache Info:              
    L1:                      16(0x10) KB                        
    L2:                      8192(0x2000) KB                    
  Chip ID:                 26287(0x66af)                      
  ASIC Revision:           1(0x1)                             
  Cacheline Size:          64(0x40)                           
  Max Clock Freq. (MHz):   1801                               
  BDFID:                   4608                               
  Internal Node ID:        1                                  
  Compute Unit:            60                                 
  SIMDs per CU:            4                                  
  Shader Engines:          4                                  
  Shader Arrs. per Eng.:   1                                  
  WatchPts on Addr. Ranges:4                                  
  Coherent Host Access:    FALSE                              
  Memory Properties:       
  Features:                KERNEL_DISPATCH 
  Fast F16 Operation:      TRUE                               
  Wavefront Size:          64(0x40)                           
  Workgroup Max Size:      1024(0x400)                        
  Workgroup Max Size per Dimension:
    x                        1024(0x400)                        
    y                        1024(0x400)                        
    z                        1024(0x400)                        
  Max Waves Per CU:        40(0x28)                           
  Max Work-item Per CU:    2560(0xa00)                        
  Grid Max Size:           4294967295(0xffffffff)             
  Grid Max Size per Dimension:
    x                        4294967295(0xffffffff)             
    y                        65535(0xffff)                      
    z                        65535(0xffff)                      
  Max fbarriers/Workgrp:   32                                 
  Packet Processor uCode:: 481                                
  SDMA engine uCode::      145                                
  IOMMU Support::          None                               
  Pool Info:               
    Pool 1                   
      Segment:                 GLOBAL; FLAGS: COARSE GRAINED      
      Size:                    16760832(0xffc000) KB              
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Recommended Granule:2048KB                             
      Alloc Alignment:         4KB                                
      Accessible by all:       FALSE                              
    Pool 2                   
      Segment:                 GLOBAL; FLAGS: EXTENDED FINE GRAINED
      Size:                    16760832(0xffc000) KB              
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Recommended Granule:2048KB                             
      Alloc Alignment:         4KB                                
      Accessible by all:       FALSE                              
    Pool 3                   
      Segment:                 GROUP                              
      Size:                    64(0x40) KB                        
      Allocatable:             FALSE                              
      Alloc Granule:           0KB                                
      Alloc Recommended Granule:0KB                                
      Alloc Alignment:         0KB                                
      Accessible by all:       FALSE                              
  ISA Info:                
    ISA 1                    
      Name:                    amdgcn-amd-amdhsa--gfx906:sramecc+:xnack-
      Machine Models:          HSA_MACHINE_MODEL_LARGE            
      Profiles:                HSA_PROFILE_BASE                   
      Default Rounding Mode:   NEAR                               
      Default Rounding Mode:   NEAR                               
      Fast f16:                TRUE                               
      Workgroup Max Size:      1024(0x400)                        
      Workgroup Max Size per Dimension:
        x                        1024(0x400)                        
        y                        1024(0x400)                        
        z                        1024(0x400)                        
      Grid Max Size:           4294967295(0xffffffff)             
      Grid Max Size per Dimension:
        x                        2147483647(0x7fffffff)             
        y                        65535(0xffff)                      
        z                        65535(0xffff)                      
      FBarrier Max Size:       32                                 
    ISA 2                    
      Name:                    amdgcn-amd-amdhsa--gfx9-generic:xnack-
      Machine Models:          HSA_MACHINE_MODEL_LARGE            
      Profiles:                HSA_PROFILE_BASE                   
      Default Rounding Mode:   NEAR                               
      Default Rounding Mode:   NEAR                               
      Fast f16:                TRUE                               
      Workgroup Max Size:      1024(0x400)                        
      Workgroup Max Size per Dimension:
        x                        1024(0x400)                        
        y                        1024(0x400)                        
        z                        1024(0x400)                        
      Grid Max Size:           4294967295(0xffffffff)             
      Grid Max Size per Dimension:
        x                        2147483647(0x7fffffff)             
        y                        65535(0xffff)                      
        z                        65535(0xffff)                      
      FBarrier Max Size:       32                                 
*** Done ***
```

### Additional Information

Script output:
```
reading tuning results from tuned_gemm0.csv
could not open tuned_gemm0.csv for reading tuning results
missing op_signature, returning null ResultEntry for GemmTunableOp_Half_NT,nt_65536_32768_384_ld_65536_32768_65536
finding fastest for GemmTunableOp_Half_NT(nt_65536_32768_384_ld_65536_32768_65536) out of 546 candidates
```

### PR DESCRIPTION
**Pull request by adityas-amd**
## Summary
  Fixes https://github.com/ROCm/TheRock/issues/5553. Offline TunableOp tuning (`tune_gemm_in_file`) could silently tune the wrong GEMM shape. `_process_single_offline_gemm` decided whether an entry described a sub-matrix (padded leading dimensions) with:
```python
  if all(item in [n, m, k] for item in [lda, ldb, ldc]):
      subMatrix = False
```

  This only checks membership in {n, m, k}. A padded leading dimension can legitimately equal one of n/m/k; when it does, the GEMM is misclassified as tight, the requested leading dimensions are dropped, and contiguous matrices are built instead — tuning a different shape than requested.

Example (from the linked issue): tuning GemmTunableOp_Half_NT,nt_65536_32768_384_ld_65536_65536_65536 (where lda == n == 65536) was silently rewritten to  ...ld_65536_32768_65536 (lda collapsed to m == 32768).

  ## Fix

  Detect the sub-matrix case by comparing each leading dimension against its
  expected contiguous value for the transpose layout:
```
  lda_tight = m if transA else k
  ldb_tight = k if transB else n
  ldc_tight = n
  subMatrix = not (lda == lda_tight and ldb == ldb_tight and ldc == ldc_tight)
```
  The change is strictly one-directional (only ever moves cases tight → sub, never
  the reverse), so genuinely-tight GEMMs are unaffected. Verified that all four
  transpose layouts (NN/NT/TN/TT) still classify as tight.

  ## Test plan

  Adds test_mm_submatrix_leading_dim_alias_offline_tunableop, which records and
  offline-tunes an NT sub-matrix whose lda aliases n (nt_6_4_2_ld_6_6_6) and
  asserts the tuned param signature matches the recorded one.

  Verified on gfx906 (ROCm):
  - New test: passes with the fix, fails without it (tuned entry drifts to
  nt_6_4_2_ld_6_4_6).
  - Existing offline tunableop tests still pass:
  test_mm_submatrix_offline_tunableop, test_matmul_offline_tunableop,
  test_offline_tuning_append_to_existing_file_tunableop,
  test_gemm_bias_offline_tunableop.

### CODE REVIEW

**Review by jeffdaily — APPROVED** [review]

**Comment by jeffdaily** [comment]
@pytorchbot merge -i

Ignoring unrelated lint failure.
